### PR TITLE
Typehint state.py

### PR DESCRIPTION
**Pull request by StockerMC**
## Summary
This pull request typehints state.py

The other files that were edited were edited for typehinting state.py

## Checklist

- [ ] If code changes were made then they have been tested.
    - [ ] I have updated the documentation to reflect the changes.
- [ ] This PR fixes an issue.
- [ ] This PR adds something new (e.g. new method or parameters).
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, ...)

### CODE REVIEW

File: discord/state.py [hunk 1]
```diff
@@ -56,26 +54,43 @@
...
-
-    def add_members(self, members):
+    def __init__(self, guild_id: int, loop: asyncio.AbstractEventLoop, resolver: Callable[[int], Any], *, cache: bool = True) -> None:
+        self.guild_id: int = guild_id
+        self.resolver: Callable[[int], Any] = resolver
+        self.loop: asyncio.AbstractEventLoop = loop
+        self.cache: bool = cache
+        self.nonce: str = os.urandom(16).hex()
+        self.buffer: List[Member] = []
+        self.waiters: List[asyncio.Future] = []
```
Comment (Rapptz) [inline discord/state.py:91]:
```suggestion
        self.waiters: List[asyncio.Future[List[Member]]] = []
```

File: discord/state.py [hunk 2]
```diff
@@ -87,43 +102,48 @@ def add_members(self, members):
                 if existing is None or existing.joined_at is None:
                     guild._add_member(member)
 
-    async def wait(self):
+    async def wait(self) -> None:
```
Comment (Rapptz) [inline discord/state.py:105]:
```suggestion
    async def wait(self) -> List[Member]:
```

File: discord/state.py [hunk 3]
```diff
@@ -87,43 +102,48 @@ def add_members(self, members):
...
         future = self.loop.create_future()
         self.waiters.append(future)
         try:
             return await future
         finally:
             self.waiters.remove(future)
 
-    def get_future(self):
+    def get_future(self) -> asyncio.Future:
```
Comment (Rapptz) [inline discord/state.py:113]:
```suggestion
    def get_future(self) -> asyncio.Future[List[Member]]:
```

**Review by Rapptz — CHANGES_REQUESTED** [review]

File: discord/state.py [hunk 4]
```diff
@@ -1311,32 +1353,32 @@ def get_channel(self, id):
...
-        self._ready_task = None
-        self.shard_ids = ()
-        self.shards_launched = asyncio.Event()
+        self.shard_ids: Union[List[int], range] = []
+        self.shards_launched: asyncio.Event = asyncio.Event()
 
-    def _update_message_references(self):
-        for msg in self._messages:
+    def _update_message_references(self) -> None:
+        for msg in self._messages or []:
```
Comment (Rapptz) [inline discord/state.py:1366]: When this call happens `self._messages` is never `None` since it's in an `if` block.

File: discord/state.py [hunk 5]
```diff
@@ -1216,32 +1256,34 @@ def parse_stage_instance_delete(self, data):
...
 
-            member, before, after = guild._update_voice_state(data, channel_id)
+            member, before, after = guild._update_voice_state(data, channel_id) # type: ignore
             if member is not None:
                 if flags.voice:
                     if channel_id is None and flags._voice_only and member.id != self_id:
-                        # Only remove from cache iff we only have the voice flag enabled
-                        guild._remove_member(member)
+                        # Only remove from cache if we only have the voice flag enabled
+                        # Member doesn't have
```
Comment (Rapptz) [inline discord/state.py:1277]: This seems incomplete.

File: discord/state.py [hunk 6]
```diff
@@ -561,16 +592,17 @@ def parse_message_update(self, data):
...
         emoji = PartialEmoji.with_state(self, id=emoji_id, animated=emoji.get('animated', False), name=emoji['name'])
         raw = RawReactionActionEvent(data, emoji, 'REACTION_ADD')
 
         member_data = data.get('member')
         if member_data:
-            guild = self._get_guild(raw.guild_id)
-            raw.member = Member(data=member_data, guild=guild, state=self)
+            # the guild won't be None here
+            guild = self._get_guild(raw.guild_id) # type: ignore
+            raw.member = Member(data=member_data, guild=guild, state=self) # type: ignore
```
Comment (Rapptz) [inline discord/state.py:605]: This probably isn't true actually.

File: discord/state.py [hunk 7]
```diff
@@ -661,41 +693,44 @@ def parse_presence_update(self, data):
...
             old_channel = copy.copy(channel)
-            channel._update_group(data)
+            # the channel is a GroupChannel
+            channel._update_group(data) # type: ignore
             self.dispatch('private_channel_update', old_channel, channel)
             return
 
         guild_id = utils._get_as_snowflake(data, 'guild_id')
-        guild = self._get_guild(guild_id)
+        guild = self._get_guild(guild_id) # type: ignore
```
Comment (Rapptz) [inline discord/state.py:733]: I assume this was because `guild_id` is `Optional[int]`. The signature should be updated instead.

**Review by Rapptz — CHANGES_REQUESTED** [review]

File: discord/state.py (hunk 6 above)
Comment (StockerMC) [reply discord/state.py:605]: `guild_id` can be None if there's a member object provided?

File: discord/state.py (hunk 7 above)
Comment (StockerMC) [reply discord/state.py:733]: Do you mean typehinting `guild_id` as `int` and `# type: ignore`'ing that line, or changing the signature of `_get_as_snowflake`? I don't see how `_get_as_snowflake`'s signature could be changed if that was the case.

File: discord/state.py (hunk 6 above)
Comment (Rapptz) [reply discord/state.py:605]: `guild` can be `None` (e.g. guild intent is off)

File: discord/state.py (hunk 7 above)
Comment (Rapptz) [reply discord/state.py:734]: I'm talking about `_get_guild`'s signature.

Comment (StockerMC) [reply discord/state.py:734]: Alright. Should the signature for all `ConnectionState._get_x` methods that take an id be changed for the id to be `Optional[int]`?

Comment (Rapptz) [reply discord/state.py:736]: Sure.

**Review by Rapptz — APPROVED** [review]
Thanks.

**Pull request merged**